apps/passport-client/components/modals/AccountImportModal.tsx

```diff
@@ -0,0 +1,253 @@
+import { EmailPCDTypeName } from "@pcd/email-pcd";
+import { deserializeStorage } from "@pcd/passport-interface";
+import { Spacer } from "@pcd/passport-ui";
+import { PCDCollection } from "@pcd/pcd-collection";
+import { PCD } from "@pcd/pcd-types";
+import { SemaphoreGroupPCDTypeName } from "@pcd/semaphore-group-pcd";
+import { SemaphoreIdentityPCDTypeName } from "@pcd/semaphore-identity-pcd";
+import { useCallback, useEffect, useState } from "react";
+import styled from "styled-components";
+import { useFilePicker } from "use-file-picker";
+import { useDispatch, useModal, usePCDCollection } from "../../src/appHooks";
+import { getPackages } from "../../src/pcdPackages";
+import { Button } from "../core";
+
+// There are three main UI states that can occur after a user selects a file
+// to import.
+type ImportState =
+  | { state: "ready" }
+  // The selected file is valid, and the user can decide whether to import it.
+  | {
+      state: "valid-file-selected";
+      collection: PCDCollection;
+      pcdsToMergeIds: Set<PCD["id"]>;
+    }
+  // The import has been carried out, and `added` is the number of PCDs
+  // imported.
+  | { state: "import-complete" }
+  // The selected file is not valid.
+  | { state: "invalid-file" };
+
+export function AccountImportModal() {
+  const { openFilePicker, filesContent } = useFilePicker({
+    accept: ".json",
+    multiple: false,
+    onFilesSelected: () => {
+      setImportState({ state: "ready" });
+    }
+  });
+
+  const modal = useModal();
+  const pcdCollection = usePCDCollection();
+  const dispatch = useDispatch();
+
+  const [importState, setImportState] = useState<ImportState>({
+    state: "ready"
+  });
+
+  // Called when a valid file has been selected, and the user chooses to import
+  // PCDs from it.
+  const importPCDs = useCallback(() => {
+    // Should never happen, but makes TypeScript happy that we checked for it
+    if (importState.state !== "valid-file-selected") return;
+
+    dispatch({
+      type: "merge-import",
+      collection: importState.collection,
+      pcdsToMergeIds: importState.pcdsToMergeIds
+    });
+
+    setImportState({ state: "import-complete" });
+  }, [dispatch, importState]);
+
+  // Responds to the user having selected a file to import, or to changes in
+  // the user's current PCD collection.
```

Comment:
Per what Ivan taught me about React, I think this will respond if the actual collection changes (by reference) not if the collection's contents change.  I don't think that's a big deal, but it slightly broadens the cases where the dry-run and final results could differ.

Reply:
This is true in general but not in the case of `usePCDCollection` which uses a hook to cause the reference to change when an event is emitted.

Reply:
I don't think that's quite true.  usePCDCollection forces hidden state to change (used to be a hash, is now just a placeholder object), which should ensure that React re-renders.  However the actual reference to the PCDCollection doesn't change, so I don't think it would trigger any other dependency lists.  @ichub and I did discuss the fact that such an approach isn't quite perfect, but decided not to try to address it until the broader revisit of UI framework.  I think that's a fine decision to make here too, but just want to make sure we're on the same page about how things work.

Reply:
I did see it work in practice, by manually tweaking a feed to delete some PCDs without re-creating them, I was able to cause the calculation to change and prompt a re-render. But this might be confounded by sync genuinely changing the reference.

Reply:
Download definitely changes the reference.  I wouldn't expect feed issuance or upload to do so, but I'd have to walk through some code and logs to fully understand it.  Probably okay to leave this in the category of "our UI and state management is complex and flaky" to fix later.